PWGJE/TableProducer/emcalClusterHadronicCorrectionTask.cxx

```diff
@@ -0,0 +1,306 @@
+          EclusterAll1 = subtractM02ClusterEnergy(cluster.m02(), EclusterAll1, Nmatches, totalTrkP, fHadCorralltrks1, UseFraction1);
+        } else {
+          // Default energy subtraction (100% and 70%)
+          Ecluster1 = subtractClusterEnergy(Ecluster1, closestTrkP, fHadCorr1, Nmatches, UseFraction1);
+          EclusterAll1 = subtractClusterEnergy(EclusterAll1, totalTrkP, fHadCorralltrks1, Nmatches, UseFraction1);
+        }
+
+        if (UseM02SubtractionScheme2) {
+          // Do M02-based correction if enabled
+          Ecluster2 = subtractM02ClusterEnergy(cluster.m02(), Ecluster2, Nmatches, closestTrkP, fHadCorr2, UseFraction2);
+          EclusterAll2 = subtractM02ClusterEnergy(cluster.m02(), EclusterAll2, Nmatches, totalTrkP, fHadCorralltrks2, UseFraction2);
+        } else {
+          // Default energy subtraction (100% and 70%)
+          Ecluster2 = subtractClusterEnergy(Ecluster2, closestTrkP, fHadCorr2, Nmatches, UseFraction2);
+          EclusterAll2 = subtractClusterEnergy(EclusterAll2, totalTrkP, fHadCorralltrks2, Nmatches, UseFraction2);
+        }
+      }
+      registry.fill(HIST("h_Ecluster1"), Ecluster1);
+      registry.fill(HIST("h_Ecluster2"), Ecluster2);
+      registry.fill(HIST("h_EclusterAll1"), EclusterAll1);
+      registry.fill(HIST("h_EclusterAll2"), EclusterAll2);
+      registry.fill(HIST("h2_ClsEvsEcluster1"), cluster.energy(), Ecluster1);
+      registry.fill(HIST("h2_ClsEvsEcluster2"), cluster.energy(), Ecluster2);
+      registry.fill(HIST("h2_ClsEvsEclusterAll1"), cluster.energy(), EclusterAll1);
+      registry.fill(HIST("h2_ClsEvsEclusterAll2"), cluster.energy(), EclusterAll2);
+
+      // Fill the table with all four corrected energies
+      hadroniccorrectedclusters(Ecluster1, Ecluster2, EclusterAll1, EclusterAll2);
+
+    } // End of cluster loop
+  } // process function ends
+
+  // Helper function to prevent negative energy subtraction
+  double subtractClusterEnergy(double Ecluster, double mom, double correctionFactor, int Nmatches, bool UseFraction) {
+    double Ecorr = Ecluster;
+
+    // if (UseConstantSubtractionValue) {
+    if(!UseFraction) {
+      Ecorr = Ecluster- fConstantSubtractionValue* Nmatches;   // Use constant value for MIP-subtraction (regardless of the cluster-shape)
+    } else {
+      Ecorr = Ecluster - correctionFactor * mom;               // Fractional momentum subtraction
+    }
+    return (Ecorr < 0) ? 0 : Ecorr;
+  }
+
+  // Helper function for M02-based energy subtraction (based on cluster-shape)
+  double subtractM02ClusterEnergy(double m02, double Ecluster, int Nmatches, double totalTrkP, double correctionFactor, bool UseFraction) {
+    double Ecorr = Ecluster;
+
+    // For M02 in the single photon region, the signal is primarily: Single photons, single electrons, single MIPs
+    if (m02 > 0.1 && m02 < 0.4) {   //circular clusters(electron/photon)
+      Ecorr = Ecluster;  // Single electron, single MIP
+    } else if (m02 > 0.4) {
```

Comment:
Normally (I think) if M02 is bigger than 0.4 and the number of matches is >1, there should be nothing subtracted. Could you explain why you changed the logic from the function we had in Run2?